### ISSUE TITLE
ESLint in the editor - many/most errors are not highlighted, without or with a custom config

### ISSUE THREAD

**Issue by darkred** [Tampermonkey/tampermonkey#865]
**The issue**:  
- without using a custom ESLint config in TM Settings|Editor|`Custom Linter Config` textarea, 
many/most errors covered by ESLint's "recommended" rules are not highlighted in the editor, 
- when using a config file in 'Custom Linter Config' textarea, 
many/most errors covered by your config file are not highlighted.

To reproduce the former, just copy/paste various incorrect code examples from almost any "recommended" rule from https://eslint.org/docs/rules/ to a new userscript tab.

&nbsp;

**Expected behavior**: 
- when not using some custom ESLint config file, the "recommended" rules to be enabled by default.
- when using a custom ESLint config file, it be fully applied. 

&nbsp;

---

As an introduction to ESLint I quote from its site:
> ESLint is designed to be completely configurable, meaning you can turn off every rule and run only with basic syntax validation, or mix and match the bundled rules and your custom rules to make ESLint perfect for your project.

There are two primary ways to [configure](https://eslint.org/docs/user-guide/configuring)  ESLint: 
- configuration comments (inline) 
- configuration files (JavaScript, JSON or YAML).

For reference, the list of ESLint rules is here: https://eslint.org/docs/rules/
A large group of them is "recommended", and lots of them are "fixable" (autom. by ESLint itself).

<details>
<summary>An example of custom config:</summary>

```js
module.exports = {
	"env": {
		"browser": true,
		"es6": true,
		"greasemonkey": true,
		"jquery": true,
	},
	"extends": "eslint:recommended",
	// "extends": "eslint:all",
	"parserOptions": {  
		"ecmaVersion": 11,
		"sourceType": "script",
		"ecmaFeatures": {
			"globalReturn ": true,
			"impliedStrict": true,
		},
	},
	"rules": {
		"eqeqeq": "warn",
		"indent": ["warn","tab", { "ignoreComments": true, "SwitchCase": 1 } ],
		"linebreak-style": ["warn","unix"],
		"no-alert": "error",
		"no-console": "warn",
	}
};
```
</details>

---

Some notes:

- The switch in Tampermonkey's editor from JSHint to ESLint has been done since TM 4.6 (2018-05-07) <sup>[changelog](https://www.tampermonkey.net/changelog.php?ext=dhdg&show=dhdg&more=true)</sup>. Thank you, @derjanb for that.
- I had initially reported this issue [in the TM forum](https://forum.tampermonkey.net/viewtopic.php?f=26&t=2453) over 1.5 year ago, but because since the forum still offline, I thought of also reporting it here.
- If I recall correctly, @derjanb had suggested to me in the TM forum thread, 
to make a config 'Custom Linter Config' in which I should manually all "recommended" rules.
But, that's difficult to keep, because there are a lot of rules to include, plus not all use the same syntax to enable. Additionally, the "recommended" rules are not some 'static' group - in fact more rules are added in every new ESLint version.
- Using latest TM 4.10.6109 beta in Chrome 79.

**Comment by derjanb** [comment]
This now should work as expected...

```JSON
{
	"env": {
		"browser": true,
		"es6": true,
		"greasemonkey": true,
		"jquery": true
	},
	"extends": "eslint:recommended",
	"parserOptions": {  
		"ecmaVersion": 11,
		"sourceType": "script",
		"ecmaFeatures": {
			"globalReturn ": true,
			"impliedStrict": true
		}
	},
	"rules": {
		"eqeqeq": "warn",
		"indent": ["warn","tab", { "ignoreComments": true, "SwitchCase": 1 } ],
		"linebreak-style": ["warn","unix"],
		"no-alert": "error",
		"no-console": "warn"
	}
}
```

**Comment by darkred** [comment]
Thank you so much! I confirm it works great! Awesome!

Just two suggestions, please:

1. The tooltip as you mouseover an error, to contain the relevant ESLint rule name,  
e.g. instead of plain 
    ```
    Unexpected console statement.
    ```
    to become 
    ```
    eslint(no-console)    Unexpected console statement.
    ```
2. Add some kind of panel/list to list all ESLint errors/warnings.

<details>
 <summary>Here are screenshots of SublimeLinter for reference:</summary>

![2020-02-22_215323~](https://user-images.githubusercontent.com/723651/75098422-a5adcf00-55be-11ea-9f19-0b8ad56d8962.jpg)
![2020-02-22_215553](https://user-images.githubusercontent.com/723651/75098378-23bda600-55be-11ea-9202-158f5a3751d9.jpg)
</details>

**Comment by darkred** [comment]
Thanks a lot for adding the tooltip with the related rule name!

---

I've noticed that many errors are not highlighted:
(using TM beta 4.10.6112 in Chrome 81)
e.g. in the following code example:

```js
var a;       // eslint:no-unused-vars  'a' is defined but never used.
var a;       // eslint:no-redeclare    'a' is already defined.
b            // eslint:no-undef        'b' is not defined.
```
using the config from your comment, only the 1st warning is highlighted, even though all 3 rules are 'recommended'.
Also, that 1st highlight should be an error, not a highlight, because that (recommended) rule is not explicitly defined in your config as error (`"no-unused-vars": "error"`)

**Comment by derjanb** [comment]
Fixed at TM BETA 4.11.6116

**Comment by darkred** [comment]
@derjanb Thank you!

I've noticed the following with the new version:

if I add the ESLint config from your [comment](https://github.com/Tampermonkey/tampermonkey/issues/865#issuecomment-589987608)
and enter my test code from above,
then only the 1nd + 3rd issues appear as Errors (red) - **the 2nd appears as Warning (yellow)**
although the `no-redeclare` rule is not explicitly set as "Warn" in the config:

![2020-07-19_193559](https://user-images.githubusercontent.com/723651/87880036-1be02180-c9f7-11ea-96a5-afc73d8ca3bd.jpg)

**Comment by derjanb** [comment]
This is because

no-unused-vars is a "problem"
https://github.com/eslint/eslint/blob/8108f49f9fa0c2de80b3b66c847551beff585951/lib/rules/no-unused-vars.js#L20

while no-redeclare is a "suggestion".
https://github.com/eslint/eslint/blob/8108f49f9fa0c2de80b3b66c847551beff585951/lib/rules/no-redeclare.js#L20

**Comment by darkred** [comment]
I'm aware of that, 
but in my .eslintrc.js configuration file that I use (with Sublime Text/VSCode) 
I may set a rule as a warning or error, regardless of whether the the rule is a problem or a suggestion. 
https://eslint.org/docs/user-guide/configuring#configuring-rules

e.g. the `no-unused-vars` rule as warning: (ST screenshot)
![Sublime Text screenshot](https://user-images.githubusercontent.com/723651/87992292-9f316e00-caf0-11ea-91c3-517cffc92e9d.jpg)


Can't this be possible within Tampermonkey using the following config?

```json
{
	"env": {
		"browser": true,
		"es6": true,
		"greasemonkey": true,
		"jquery": true
	},
	"extends": "eslint:recommended",
	"parserOptions": {  
		"ecmaVersion": 11,
		"sourceType": "script",
		"ecmaFeatures": {
			"globalReturn ": true,
			"impliedStrict": true
		}
	},
	"rules": {
		"no-unused-vars": ["warn", {"vars": "all", "args": "after-used"}],
}
```

**Comment by darkred** [comment]
*(Please ignore my last comment)*

What I've noticed is that,
if I use in Tampermonkey a config with an `"rules"` key with an empty value `{}`:

```json
{
    "env": {
        "browser": true,
        "es6": true,
        "greasemonkey": true,
        "jquery": true
    },
    "extends": "eslint:recommended",
    "parserOptions": {
        "ecmaVersion": 11,
        "sourceType": "script",
        "ecmaFeatures": {
            "globalReturn ": true,
            "impliedStrict": true
        }
    },
    "rules": {}
}
```
then 2nd rule is highlighted as a warning:
![](https://user-images.githubusercontent.com/723651/87880036-1be02180-c9f7-11ea-96a5-afc73d8ca3bd.jpg)

&nbsp;

but if I use the same config as a .eslintrc.js file in ESLint CLI (i.e. outside Tampermonkey) :

```json
module.exports = {
    "env": {
        "browser": true,
        "es6": true,
        "greasemonkey": true,
        "jquery": true

    },
    "extends": "eslint:recommended",
    "parserOptions": {
        "ecmaVersion": 11,
        "sourceType": "script",
        "ecmaFeatures": {
            "globalReturn ": true,
            "impliedStrict": true
        }
    },
    "rules": {}
}
```

then the 2nd problem is an also highlighted as an error:
![2020-07-21_130633](https://user-images.githubusercontent.com/723651/88041427-0c75eb00-cb53-11ea-8cf9-4864d8f7ebba.jpg)

&nbsp;

In other words, all problems in ESLint CLI are highlighted as errors (regardless of whether the relevant rule is a problem or a suggestion) unless set they are explicitly set as warnings.

What you are referring to as `suggestion/problem` distinguish is the [`--fix-type`](https://eslint.org/docs/user-guide/command-line-interface#fix-type) CLI option 
which *allows to to specify the type of fixes to apply when using either `--fix` or `--fix-dry-run`*, 
and which is irrelevant to how the problems are highlighted (warning/error), i.e. how rules are configured (off/warn/error) [link](https://eslint.org/docs/user-guide/configuring#configuring-rules) . 

So, can we have the same behavior in Tampermonkey as in CLI ?

**Comment by darkred** [comment]
Thank you for your overall work regarding this!

Using latest TM v4.12.6123 beta in Chrome 87.

I'd like to report a thing I noticed regarding the `indent` rule:
although I have sent in my config to ignore comments,
I wrongly get `eslint:indent  - Expected indentation of 1 tabs but found 0.`
when just opening a new(empty) userscript tab:
screenshot:
![2020-12-28_231419](https://user-images.githubusercontent.com/723651/103243453-8f04a100-4962-11eb-81a9-29cd25690c23.jpg)

The relevant indent rule, used inline would be:
`/* eslint indent: [ "warn", "tab", { "SwitchCase": 1, "ignoreComments": true}] */`
[ESLint online demo/test page](https://eslint.org/demo)


The ESLint config that I use inside Tampermonkey:

<details>
<summary>Click to expand:</summary>

```json
{
    "env": {
        "browser": true,
        "es6": true,
        "greasemonkey": true,
        "jquery": true
    },
    "extends": "eslint:recommended",
    "globals": {
        "GM": "writeable",
        "GM_addStyle": "writeable",
        "GM_addValueChangeListener": "writeable",
        "GM_deleteValue": "writeable",
        "GM_download": "writeable",
        "GM_getResourceText": "writeable",
        "GM_getResourceURL": "writeable",
        "GM_getTab": "writeable",
        "GM_getTabs": "writeable",
        "GM_getValue": "writeable",
        "GM_info": "writeable",
        "GM_listValues": "writeable",
        "GM_log": "writeable",
        "GM_notification": "writeable",
        "GM_openInTab": "writeable",
        "GM_registerMenuCommand": "writeable",
        "GM_removeValueChangeListener": "writeable",
        "GM_saveTab": "writeable",
        "GM_setClipboard": "writeable",
        "GM_setTab": "writeable",
        "GM_setValue": "writeable",
        "GM_unregisterMenuCommand": "writeable",
        "GM_xmlhttpRequest": "writeable",
        "TM_info": "writeable",
        "uneval": "writeable",
        "unsafeWindow": "writeable"
    },
    "parserOptions": {
        "ecmaFeatures": {
            "globalReturn ": true,
            "impliedStrict": true
        },
        "ecmaVersion": 11,
        "sourceType": "script"
    },
    "reportUnusedDisableDirectives": true,
    "rules": {
        "complexity": [
            "warn",
            20
        ],
        "eqeqeq": "warn",
        "func-style": "off",
        "indent": [
            "warn",
            "tab",
            {
                "SwitchCase": 1,
                "ignoreComments": true
            }
        ],
        "linebreak-style": [
            "warn",
            "unix"
        ],
        "max-len": "off",
        "max-statements-per-line": "off",
        "new-cap": "off",
        "no-alert": "warn",
        "no-console": "warn",
        "no-extra-semi": "warn",
        "no-inline-comments": "off",
        "no-magic-numbers": "off",
        "no-misleading-character-class": "warn",
        "no-mixed-spaces-and-tabs": "warn",
        "no-multiple-empty-lines": "off",
        "no-tabs": "off",
        "no-unused-labels": "warn",
        "no-unused-vars": [
            "warn",
            {
                "args": "after-used",
                "vars": "all"
            }
        ],
        "no-useless-escape": "warn",
        "padded-blocks": "off",
        "quotes": [
            "warn",
            "single",
            {
                "allowTemplateLiterals": true
            }
        ],
        "require-jsdoc": "off",
        "require-unicode-regexp": "off",
        "semi": [
            "warn",
            "always"
        ],
        "space-before-function-paren": "off",
        "unicode-bom": [
            "warn",
            "never"
        ]
    }
}
```

</details>

**Comment by derjanb** [comment]
> The ESLint config that I use inside Tampermonkey:
> Using latest TM v4.12.6123 beta in Chrome 87.

Should be fixed at 4.12.6124. I'm not sure the Chrome Webstore team will be able to review this version this year, but in the meantime, you can download it from [here](https://www.tampermonkey.net/crx/tampermonkey_beta_4_12_6124.crx).

**Comment by darkred** [comment]
Yes, it's fixed (and available in CWS). Thanks a lot for the quick response!
And, happy new year!

**Comment by Luckz** [comment]
Setting `ecmaVersion` to more than is currently available has horrendous results, so I find it safer to leave it undefined and instead do:

```
{
    "env": {
        "browser": true,
        "es6": true,
        "es2020": true,
        "es2021": true,
        "greasemonkey": true,
        "jquery": true
    },
```

While `es2021` has no effect in TM, it doesn't seem to cause any harm either.

**Comment by nilslindemann** [comment]
The following disables the nagging about missing version numbers and descriptions in script headers. I do not need this in my innocent local scripts.

```
{
    // ...

    "plugins": [
        "userscripts"
    ],
    "rules": {
        "userscripts/require-version": false,
        "userscripts/require-description": false
    }
}
```
